Mathlib/Tactic/CC.lean

```diff
@@ -0,0 +1,325 @@
+  s.eqcOfCore e e []
+#align cc_state.eqc_of Mathlib.Tactic.CC.CCState.eqcOf
+
+/-- The equivalence class of `e`. -/
+def eqcSize (s : CCState) (e : Expr) : Nat :=
+  s.eqcOf e |>.length
+#align cc_state.eqc_size Mathlib.Tactic.CC.CCState.eqcSize
+
+/-- Continue to apply `f` using following expressions in the equivalence class of `first` to `a`
+until `first` is found. -/
+partial def foldEqcCore {α} (s : CCState) (f : α → Expr → α) (first : Expr) (c : Expr) (a : α) :
+    α :=
+  let new_a := f a c
+  let next := s.next c
+  if next == first then new_a else foldEqcCore s f first next new_a
+#align cc_state.fold_eqc_core Mathlib.Tactic.CC.CCState.foldEqcCore
+
+/-- Fold the function of `f` over the equivalence class of `e`. -/
+def foldEqc {α} (s : CCState) (e : Expr) (a : α) (f : α → Expr → α) : α :=
+  foldEqcCore s f e e a
+#align cc_state.fold_eqc Mathlib.Tactic.CC.CCState.foldEqc
+
+/-- Fold the monadic function of `f` over the equivalence class of `e`. -/
+def foldEqcM {α} {m : Type → Type} [Monad m] (s : CCState) (e : Expr) (a : α)
+    (f : α → Expr → m α) : m α :=
+  foldEqc s e (return a) fun act e => do
+    let a ← act
+    f a e
+#align cc_state.mfold_eqc Mathlib.Tactic.CC.CCState.foldEqcM
+
+end CCState
+
+/--
+The congruence closure tactic `cc` tries to solve the goal by chaining
+equalities from context and applying congruence (i.e. if `a = b`, then `f a = f b`).
+It is a finishing tactic, i.e. it is meant to close
+the current goal, not to make some inconclusive progress.
+A mostly trivial example would be:
+
+```lean
+example (a b c : ℕ) (f : ℕ → ℕ) (h: a = b) (h' : b = c) : f a = f c := by
+  cc
+```
+
+As an example requiring some thinking to do by hand, consider:
+
+```lean
+example (f : ℕ → ℕ) (x : ℕ)
+    (H1 : f (f (f x)) = x) (H2 : f (f (f (f (f x)))) = x) :
+    f x = x := by
+  cc
+```
+
+The tactic works by building an equality matching graph. It's a graph where
+the vertices are terms and they are linked by edges if they are known to
+be equal. Once you've added all the equalities in your context, you take
+the transitive closure of the graph and, for each connected component
+(i.e. equivalence class) you can elect a term that will represent the
+whole class and store proofs that the other elements are equal to it.
+You then take the transitive closure of these equalities under the
+congruence lemmas.
+
+The `cc` implementation in Lean does a few more tricks: for example it
+derives `a = b` from `Nat.succ a = Nat.succ b`, and `Nat.succ a != Nat.zero` for any `a`.
+
+* The starting reference point is Nelson, Oppen, [Fast decision procedures based on congruence
+closure](http://www.cs.colorado.edu/~bec/courses/csci5535-s09/reading/nelson-oppen-congruence.pdf),
+Journal of the ACM (1980)
+
+* The congruence lemmas for dependent type theory as used in Lean are described in
+[Congruence closure in intensional type theory](https://leanprover.github.io/papers/congr.pdf)
+(de Moura, Selsam IJCAR 2016). -/
```

Comment:
I think this docstring should be used only on the `cc` tactic elaborator itself (and then perhaps copied via `@[inherit_doc cc]`). Otherwise someone will modify one but not the other.

```suggestion
/-- Solve the metavariable using the congruence closure tactic.

See the docstring on the `cc` tactic for a full explanation of how the tactic works.
-/
```

Reply:
An alternative is to keep the explanation of the algorithm (everything from "The tactic works by ..." onwards) here, removing it from the `cc` tactic docstring. The tactic user doesn't need to know too much about the implementation, after all.

Reply:
If we use the `inherit_doc` attribute on the elaborator, tactic docs can't be displayed.